### PR TITLE
improved USB serial module

### PR DESCRIPTION
**Pull request by stevstrong**
- improved Rx routine with bugfix for loosing bytes when reading very slowly or very fast the Rx buffer (see http://www.stm32duino.com/viewtopic.php?f=9&t=1451)
- reworked Tx routine - implemented buffered transmission, which improves significantly the Tx speed

### CODE REVIEW

**Comment by jaretburkett** [comment]
There are two PRs that will conflict. I have tested both Stevstrong's and Rasmusfk's. They both completly solve both issues in the CDC core and use two different methods to solve them. I have tested with multiple tests from images, single bytes, and readbytes into a buffer. The one by stevstrong is a little faster on my read bytes testing. Stevstrong's version averages ~316 kBps on reads and Rasmusfk's version averages ~ 278 kBps on reads. Both are improvements on the current version which is only ~242 kBps. I am running tests for 2 different image to tft programs I have, as well as the test files provided at [PJRC http://www.pjrc.com/teensy/benchmark_usb_serial_receive.html](http://www.pjrc.com/teensy/benchmark_usb_serial_receive.html). I think others should run tests on both (I am testing in a linux environment), but my tests lead me to recommend Stevstrong's version. (I am posting this on both PRs.)

**Comment by rogerclarkmelbourne** [comment]
Has this PR been superseded by https://github.com/rogerclarkmelbourne/Arduino_STM32/pull/221 ?

I'm not sure what the current status of this fix is.

Can someone let me know, so can merge the appropriate files (PR) etc

**Comment by stevstrong** [comment]
Well, I would tend to say that this PR should supersede #220 .

**Comment by stevstrong** [comment]
Ups, I didn't want these changes as PR, how to delete them from PR?

**Comment by rogerclarkmelbourne** [comment]
Stev

I'll try to manually separate out the USB and the SPI stuff and then merge it

**Comment by rogerclarkmelbourne** [comment]
I've manually merged just your USB files.

Can you check the master branch of this repo, as it should now contain your usb files

**Comment by rogerclarkmelbourne** [comment]
@stse

**Comment by stevstrong** [comment]
@Roger, here are the changes, in short:
- all functions are indeed blocking, I just corrected the comment
- removed commented out parts
- corrected both SPI single byte and buffered read functions
- optimized code size for write and transfer functions, merge into single function of both 8 bit and 16 bit operations
I have reverted the changes to easy the merging process, but you merged the PR manually so that this is now superfluous.
If you consider that these changes would be accepted for the main branch I could prepare another PR.

**Comment by rogerclarkmelbourne** [comment]
I only manually merged the USB code

I'll need to manually merge the SPI (tomorrow)

**Comment by stevstrong** [comment]
You could use [my repo](https://github.com/stevstrong/Arduino_STM32/tree/master/STM32F1/libraries/SPI/src) as reference.

**Comment by rogerclarkmelbourne** [comment]
Thanks

**Comment by rogerclarkmelbourne** [comment]
Steve,

You latest changes are not compatible with existing code

I tried to compile the ILI9341 graphic test and I get a load of errors



```
\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:436:31: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int16_t&, int)'

   SPI.dmaSend(lineBuffer, h, 0);

   

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:436:31: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);



\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp: In member function 'virtual void Adafruit_ILI9341_STM::drawFastHLine(int16_t, int16_t, int16_t, uint16_t)':

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:469:31: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int16_t&, int)'

   SPI.dmaSend(lineBuffer, w, 0);

   

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:469:31: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);



\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp: In member function 'virtual void Adafruit_ILI9341_STM::fillScreen(uint16_t)':

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:490:37: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int, int)'

   SPI.dmaSend(lineBuffer, (65535), 0);

 

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:490:37: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);



\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:491:58: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int, int)'

   SPI.dmaSend(lineBuffer, ((_width * _height) - 65535), 0);

  

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:491:58: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);



\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp: In member function 'virtual void Adafruit_ILI9341_STM::fillRect(int16_t, int16_t, int16_t, int16_t, uint16_t)':

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:521:35: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int, int)'

   SPI.dmaSend(lineBuffer, (w*h), 0);

   

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:521:35: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);


\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:524:37: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int, int)'

   SPI.dmaSend(lineBuffer, (65535), 0);

 

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:524:37: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);



\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:525:45: error: no matching function for call to 'SPIClass::dmaSend(uint16_t [320], int, int)'

   SPI.dmaSend(lineBuffer, ((w*h) - 65535), 0);

 

\STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:525:45: note: candidate is:

In file included from \STM32F1\libraries\Adafruit_ILI9341_STM\Adafruit_ILI9341_STM.cpp:12:0:

\STM32F1\libraries\SPI\src/SPI.h:286:8: note: uint8 SPIClass::dmaSend(void*, uint16)

  uint8 dmaSend(void * transmitBuf, uint16 length);



\STM32F1\libraries\SPI\src/SPI.h:286:8: note:   candidate expects 2 arguments, 3 provided

```

**Comment by stevstrong** [comment]
Hi Roger,
it seems that I removed a "feature" which I considered superfluous. In these cases the `dmaSend()` function seems to be "misused" by programming the DMA to transfer the same data for a specific number of times.

I see two solutions:
1. I revert (re-code) the removed "feature". In this case no further action is needed.
2. I would implement another function which clearly indicates that it is about a repeated write of a single data (byte/word) (ex. `dmaSendRepeat(void data, uint16 length)` rather than transferring a whole buffer of data. - the few places which are using this feature would need to be changed accordingly.

Which one shall be?

Btw, using now the "normal" functions `read/write(const void *data, uint32 length)` (w/0 DMA) gives the same performance as using the DMA, because the small wait periods between consecutive data transfers has been removed.

**Comment by rogerclarkmelbourne** [comment]
You still call

+		while ( spi_is_tx_empty(dev)==0 ); // wait Tx to be empty

So I am not sure how sending single bytes is as fast as DMA

**Comment by rogerclarkmelbourne** [comment]
Probably the best fix to the DMAsend issue is to add the replacement function and update all examples and libraries that use the old function.

I suspect It's only used in the display driver libraries.

You could also put back a stub function for DMASend, and get it to throw an error using # error Please use dmaSendRepeat

so it would be obvious how to fix compatibility issues

**Comment by stevstrong** [comment]
The TXE flag is set after loading the data into the serial output register, which happens at sending data bit b1.
Hence, sending bytes **can** be continuous (thus as fast as DMA) **IF** the software is capable to use the period between transferring bits b2 to b7 to load the next byte to be transmitted into SPI_DR register:

image

As you can see from the figure, 0xF2 is transferred right after 0xF1, without any gap.

Of course, this requires the CPU to be able to detect TXE==1 and load the next data within 6 SCK pulses, which in case of high SPI clock frequencies may not be accomplished. However, at 9MHz I could reach continuous data output (4 instructions of a F103C running at 72MHz with 2 wait states). At 36MHz, when the buffer is larger than a threshold number, the DMA is faster and thus a better choice.

**Comment by stevstrong** [comment]
> You could also put back a stub function for DMASend, and get it to throw an error using # error Please use dmaSendRepeat

Roger, as I am not expert in either C++ programming or GCC compiler, I would need more info about how exactly this stub function should look like. I am able to code a simple #error pragma, but not really more. :\

**Comment by rogerclarkmelbourne** [comment]
Hi @stevstrong 
Actually, Perhaps it is not possible to use # error as I think it will stop the compilation even if the function is not used.

I read this thread  http://stackoverflow.com/questions/4694896/generating-link-time-error-for-deprecated-functions

But I think it is all too complicated

So it is best just to comment out the old function, and to update the libraries which used it

**Comment by rogerclarkmelbourne** [comment]
Re: DMA

A lot of people use SPI faster than the recommended values i.e at DIV2 = 36Mhz, which is why DMA is important as its definitely faster than non-dma where the processor has to check for the hardware registers

Ideally we should include a non-blocking dmaTransfer function with a completion callback.

**Comment by rogerclarkmelbourne** [comment]
Can you let me know what the optimizations are

**Comment by stevstrong** [comment]
Hi Roger,

I added some of the most important changes in the comment field of the commit, but here are all the changes in detail:

**spi.c**
- line 96: speed-optimization: instead of an "if" branch within a "loop", now two loops using one "if" branch. Additionally, TXE bit checking function replaced by direct register access, which gives further speed improvement.

**SPI.cpp**
- line 163: relevant for slave mode only: removed flag "SPI_SW_SLAVE" (the slave select should always be detected by the hardware) and added "SPI_RX_ONLY" to avoid any possible collision in a multi-slave configuration (see RM0008, p734)
- lines 309,312: the single data read function now can read 16bit data
- line 324: uups, it seems to be wrong, correct should be: 
```
void SPIClass::read(uint8 *buf, uint32 len)
{
	spi_reg_map * regs = _currentSetting->spi_d->regs;
	uint8 b = (regs->DR); // clear the RX buffer in case a byte is waiting on it.
	// start sequence
	while ( (len--)>0) {
		regs->DR = 0x00FF; // " write the data item to be transmitted into the SPI_DR register (this clears the TXE flag)."
		while ( (regs->SR & SPI_SR_RXNE)==0 ) ; // wait till data is available in the Rx register
		*buf++ = (uint8)(regs->DR); // read and store the received byte
	}
}
```
I have changed this function because it is used by SdFat (beta) lib, so I though speed optimization makes sense.
- line 335: changed comment. Ups (again, sorry), it seems that I deleted unintentionally the line
```while (spi_is_tx_empty(_currentSetting->spi_d) == 0); // "5. Wait until TXE=1 ..."```
The funny thing is that the SW worked correctly so that it seems that this line is actually superfluous. By analyzing the code flow, the TXE bit is actually immediately set after writing the new value in data register, that´s why checking it gives always true, hence superfluous, BSY bit checking seems to be enough to detect end of transmission. Maybe we could remove this line from all other places, too.
- line 339: I added a function to write the same data repeatedly for a specified number of times. This functionality is used in some display libs and SdFat lib (also see next following comment below).
- lines 349-361: changed order of functions "transfer()" and "transfer16()" (cosmetic)
- lines 422, 425: regarding our conversation in previous posts above, I decided to revert the missing input parameter "increment flag" by dmaSend. In this version it is possible to use this function with or without this input parameter, thus satisfying the requirements of current implementations of some display libs. Adding a new function for repeatedly writing same data I found not optimal because it would introduce some very seldom used code. For non-DMA repeated writing I added another function with small footprint (code size), as mentioned above.
- line 441: I made clear what exactly the software is checking.

**SPI.h**
- line 118: added new  pattern for SPISettings to be used only with clock frequency as input parameter - this is used in the new SdFat beta version, so I thought it would make sense in general.
- the remaining changes are respectively related to the changes listed for "SPI.cpp".

**Comment by rogerclarkmelbourne** [comment]
Thanks

I'll pull the version locally and test it with my ILI9341 display

**Comment by rogerclarkmelbourne** [comment]
Hi Steve

It compiles, but does not work with the ILI9341 display

Data is getting set to the display, but rather than the standard test display of lines and filled shapes, I just seem to get horizontal or vertical lines and it looks generally messed up.

img_20161212_195155224 - copy
img_20161212_195158214_hdr - copy
img_20161212_195205755 - copy

**Comment by stevstrong** [comment]
Hm, it looks bad. Did you enable the DMA feature in SPI.cpp?
Unfortunately, I don't have any SPI display, I currently only test with SD card module using SdFat lib.
Seems that I should finally buy a real ILI9341 SPI display. Do you have any reference where to buy from (Aliexpress)?

**Comment by rogerclarkmelbourne** [comment]
@stevstrong 

I'm just running the normal STM32 ILI9341 display test, which includes

`#include <Adafruit_ILI9341_STM.h> // STM32 DMA Hardware-specific library`

Which uses DMA

There may be a non-dma version of the library somewhere, but I'm not sure whether it has been ported to STM32

**Comment by stevstrong** [comment]
As I wrote, I don't have any SPI driven display module at the moment, so I cannot test the display routines which uses DMA.
I ordered [one of this](https://www.aliexpress.com/item/Rasberry-pi-Raspberry-Pi-2-LCD-3-5-inch-TFT-touch-screen-Display-kit-RPI-Touch/32551712173.html), but it will take a while till arrives.
Meanwhile I continuously test with SdFat (beta) lib, which is working so far OK with enabled DMA.
Maybe there are essential differences between how SdFat and display drivers use the DMA, but I cannot determine these differencies without being able to drive a display over SPI.
Blindly, I am thinking on wrong unit8/unit16 casted data, or maybe endianess.
I would go this way: a custom test which only displays, let's say, a point on the screen. Then a line (vertical and/or horozontal), later followed maybe by a rectangle. And so slowly, step by step, one could identify the problem.

**Comment by rogerclarkmelbourne** [comment]
@stevstrong 

Thanks.

I will change the setDataSize function and see if that fixes the problem

**Comment by stevstrong** [comment]
> Does the display you ordered use a ILI9341 ?
I don't think so, but I think it is not very important to be same type, more important is to use similar functions.

**Comment by stevstrong** [comment]
I have identified (and corrected) a potential problem which could lead to conflict:
In the ILI9241 TFT lib the 8 bit data size transfer for SPI is set by:
`SPI.setDataSize (0);`.
Correct would be:
`SPI.setDataSize(DATA_SIZE_8BIT);`, meaning a non-zero parameter.

Now I have adapted in `SPI.cpp` to automatically detect the 16 bit data size instead of 8 bit data size ([see here](https://github.com/rogerclarkmelbourne/Arduino_STM32/pull/220/commits/fb823b6929e4416b0921b4ff6c9ea4fa5a186262#diff-48b5f6ab32a24d9cd0835c829b41c629R392)). This way the parameter value is handled correctly in the `dmaSend()` function, even by an erroneous input parameter.

Roger, please try this SPI version with your ILI9341 display, it should work now, I could successfully test it with another SPI driven TFT display (see [my post](http://stm32duino.com/viewtopic.php?f=9&t=1654)).

**Comment by rogerclarkmelbourne** [comment]
I just pulled to a local branch and it seems to work fine on my ILI9341 display

Can you confirm whether you have pulled all my latest changes to your repo? Because when I pulled, I noticed that files like tone and wirish shift were uploaded but boards.txt was not updated.

> Merge made by the 'recursive' strategy.
>  .gitignore                           |  2 ++
>  README.md                            |  2 +-
>  STM32F1/cores/maple/io.h             |  2 ++
>  STM32F1/cores/maple/tone.cpp         | 40 +++++++++++++++++++++---------------
>  STM32F1/cores/maple/wirish.h         |  1 +
>  STM32F1/cores/maple/wirish_shift.cpp | 25 ++++++++++++++++++++++
>  STM32F1/libraries/SPI/src/SPI.cpp    | 21 +++++++++----------
>  STM32F4/cores/maple/io.h             |  2 ++
>  STM32F4/cores/maple/wirish_shift.cpp | 25 ++++++++++++++++++++++
>  9 files changed, 92 insertions(+), 28 deletions(-)

So I don't want to merge unless your repo is totally up to date.


Thanks

Roger

**Comment by stevstrong** [comment]
Roger, I just update and synced from your master branch, checked locally `boards.txt` and it contains the latest changes. Also, ILI9341_due_STM has been removed.
No other differences are shown by my github manager for windows. So I think I am up-to-date now.

**Comment by rogerclarkmelbourne** [comment]
@stevestrong

I tried pulling this, and it does still work on my ILI9341 test board

But its slower than the current code.

Existing code gave this output from the STM32_graphictest

ILI9341 Test!
Display Power Mode: 0x0
MADCTL Mode: 0xDE
Pixel Format: 0xFC
Image Format: 0x7
Self Diagnostic: 0xDE
Benchmark                Time (microseconds)
Screen fill              170786
Text                     43084
Lines                    242454
Horiz/Vert Lines         15812
Rectangles (outline)     11567
Rectangles (filled)      355068
Circles (filled)         143378
Circles (outline)        172903
Triangles (outline)      59724
Triangles (filled)       166037
Rounded rects (outline)  59773
Rounded rects (filled)   415715
Done!

After I pulled you changed and quit the IDE and reloaded and recompiled I got these figures

ILI9341 Test!
Display Power Mode: 0x0
MADCTL Mode: 0x0
Pixel Format: 0x0
Image Format: 0xFF
Self Diagnostic: 0x0
Benchmark                Time (microseconds)
Screen fill              170800
Text                     46783
Lines                    269524
Horiz/Vert Lines         16098
Rectangles (outline)     11980
Rectangles (filled)      355169
Circles (filled)         157778
Circles (outline)        190090
Triangles (outline)      66589
Triangles (filled)       173713
Rounded rects (outline)  65082
Rounded rects (filled)   420203
Done!

So basically your code appears to be marginally slower

From what I remember, your code fixes some issues, but I thought it was going to be faster.

Was it supposed to be faster or was this just bug fixes which have to be included, but which cause the code to run slower ??

**Comment by stevstrong** [comment]
Hi Roger,
I am investigating this, it also looks like the reading of registers is somehow corrupted (display mode, pixel format, image format).

**Comment by stevstrong** [comment]
Roger, could you please post here the `SPI.cpp` version you used for testing? Thanks.

**Comment by stevstrong** [comment]
I compared the `SPI.cpp` versions from official repo (yours) and mine. The following changes seem to be relevant here:

1. My version of `SPI.setDataSize()` got 2 more code lines to temporarily disable the SPI for setting the new data size value in CR1 register. This leads to run-time increase, that's sure. As the ILI9341 TFT lib is very frequently using this function, I can imagine that the run-time increase (up to ~10%) you measured is caused by this change, specially for single pixel writes, involved by setting the display address window.
As workaround and test, try to keep your version in this case to check whether the lib still works. If yes, than the change seems to be superfluous and should be removed. Otherwise I suppose we have to live with that, or change the ILI9341 lib.
The other very frequently used function, `SPI.write()`, was not touched.

2. The function `SPI.dmaSend()` was also amended in my version to unify the 8 bit and 16 bit transfer versions in one function. This involved an addition of 2 code lines. In this case however, I don't think this would cause any significant run-time increase, because these additional lines represent maybe 0.1% of the total run-time, and this function is not called very frequently.

3. Reading data with my `SPI.transfer()` seems not to work correctly under certain conditions, as shown by your log, I have to more intensively check this. Unfortunately my SPI display does not support read operations, so I have to somehow artificially set up a test bench.
Please revert this change to check whether the readings will be correct.

Please let me know the results of your test.

**Comment by stevstrong** [comment]
Luckily I could test the SPI read/transfer/write functions with a nRF905 device.
This way I detected some new findings regarding the previously mentioned isues:

2. The line mentioned is mandatory, otherwise writing consecutive data to SPI will not work correctly.

3. It seems that the SPI hardware does not update the SPI_DR register when data reception overflow occurs. Test case:
```
SPI.write(0x55);
SPI.transfer(0xAA); -> will return the data recorded previously during writing 0x55.
```
The bugfix to this is to read the DR register before transferring the data in the transfer function.

So this is how the transfer function should actually look like:
```
uint8 SPIClass::transfer(uint8 byte) const
{
	spi_rx_reg(_currentSetting->spi_d); // read any previous data --- bugfix ---
	spi_tx_reg(_currentSetting->spi_d, byte); // Write the data item to be transmitted
	while (spi_is_tx_empty(_currentSetting->spi_d) == 0); // wait until TXE=1
	while (spi_is_busy(_currentSetting->spi_d) != 0); // wait until BSY=0
	return (uint8)spi_rx_reg(_currentSetting->spi_d); // read and return the last received data.
}
```

**Comment by rogerclarkmelbourne** [comment]
@stevstrong 

OK. I will do a local pull, and test again.

**Comment by stevstrong** [comment]
Roger, sorry that changes not relevant for SPI are going to be included here.
If there is still interest, I could generate a separate PR including only the relevant SPI changes, which would then supersede this PR.

**Comment by rogerclarkmelbourne** [comment]
I think it would be better if the changes for SPI were separated into a different PR

Having multiple different changes in one commit are a problem for testing and also for supporting the core, as rolling back a commit can remove some bugs but also removes different unrelated functionality

Where possible, I try to only fix one issue per commit. (Occasionally I make a mistake and commit 2 things but I try not to)

**Comment by stevstrong** [comment]
OK, I'll prepare a separate PR for SPI improvements only.

**Comment by rogerclarkmelbourne** [comment]
@stevstrong 
Thankyou.

**Comment by rogerclarkmelbourne** [comment]
Has this been superseded?

**Comment by stevstrong** [comment]
I'll check it later today.

**Comment by victorpv** [comment]
I think the SPI changes were added in a nother PR, but I believe not the USB ones.
On speeding up USB, I had a look at what Teensy had done with readBytes. They basically wrote an implementation of readBytes for the USB driver, rather than rely in the Stream class generic one that reads bytes 1 by 1 to a buffer.
I think we should probably do something similar, and was thinking on trying to write it, but would be good to have all the other fixes Steve wrote for the USB library in place first.

**Comment by stevstrong** [comment]
HI Roger,
I finally managed to sync up to your master branch, the remaining differences in 6 files are related now only to F4 (and .gitignore).
So it seems that all F1 SPI relevant changes are already in your master.

**Comment by stevstrong** [comment]
Just updated freshly from your master branch.

**Comment by stevstrong** [comment]
Hi Roger, it would be nice if we could somehow bring this a.s.a.p. to an end.
Because of my incompetence, this is "blocking" my master branch against other PRs.

**Comment by rogerclarkmelbourne** [comment]
Steve

I've manually updated the F1 changes in your PR in the hope that it would remove them from this merge, but they are still being listed.

I don't see any easy way to resolve this.

I think I'll need to download a zip of your fork, and then copy your STM32F4 folder to replace my local STM32F4 folder and then with the STM32F1 folder I'll need to copy the examples as I think I already did a lot of the USB stuff.

But its going to take me some time to do this.

**Comment by stevstrong** [comment]
Would be better if I close this issue?
And will try to make a separate (clean) one.

**Comment by rogerclarkmelbourne** [comment]
Leave this one open at the moment, until I have manually merged these changes

**Comment by stevstrong** [comment]
Roger, it seems that the remained changes only relate to F4, so it would be safe (for F1 too) to merge.

**Comment by rogerclarkmelbourne** [comment]
I'm not sure why this differences still exist, but I agree they are only in the F4 and the gitignore.

So just to get this done, I'll merge it...
and we can deal with the consequences later;-)
